internal/armada/server/submit.go

```diff
@@ -219,137 +245,169 @@ func (server *SubmitServer) SubmitJobs(ctx context.Context, req *api.JobSubmitRe
+	}
+	if len(jobs) != 1 {
+		return nil, status.Errorf(codes.Internal, "[cancelJobsById] error getting job with ID %s: expected exactly one result, but got %v", jobId, jobs)
+	}
+
+	result, err := server.cancelJobs(ctx, jobs[0].Queue, jobs)
+	var e *ErrNoPermission
+	if errors.As(err, &e) {
+		return nil, status.Errorf(codes.PermissionDenied, "[cancelJobsById] error canceling job with ID %s: %s", jobId, e)
+	} else if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "[cancelJobsById] error checking permissions: %s", err)
+	}
+
+	return result, nil
+}
+
+// cancels all jobs part of a particular job set and queue
+func (server *SubmitServer) cancelJobsByQueueAndSet(ctx context.Context, queue string, jobSetId string) (*api.CancellationResult, error) {
+	ids, err := server.jobRepository.GetActiveJobIds(queue, jobSetId)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "[cancelJobsBySetAndQueue] error getting job IDs: %s", err)
+	}
+
+	// Split IDs into batches and process one batch at a time
+	// To reduce the number of jobs stored in memory
+	batches := util.Batch(ids, server.cancelJobsBatchSize)
+	cancelledIds := []string{}
+	for _, batch := range batches {
+		jobs, err := server.jobRepository.GetExistingJobsByIds(batch)
 		if err != nil {
-			return nil, status.Errorf(codes.Aborted, err.Error())
+			result := &api.CancellationResult{CancelledIds: cancelledIds}
+			return result, status.Errorf(codes.Internal, "[cancelJobsBySetAndQueue] error getting jobs: %s", err)
 		}
 
-		batches := util.Batch(ids, server.cancelJobsBatchSize)
-		cancelledIds := []string{}
-		for _, batch := range batches {
-			jobs, err := server.jobRepository.GetExistingJobsByIds(batch)
-			if err != nil {
-				return &api.CancellationResult{CancelledIds: cancelledIds}, status.Errorf(
-					codes.Internal,
-					"failed to find some jobs: %v", err)
-			}
-			result, err := server.cancelJobs(ctx, request.Queue, jobs)
-			if err != nil {
-				return &api.CancellationResult{CancelledIds: cancelledIds}, status.Errorf(
-					codes.Internal,
-					"failed to cancel some jobs: %v", err)
-			}
-			cancelledIds = append(cancelledIds, result.CancelledIds...)
+		result, err := server.cancelJobs(ctx, queue, jobs)
+		var e *ErrNoPermission
+		if errors.As(err, &e) {
+			return nil, status.Errorf(codes.PermissionDenied, "[cancelJobsBySetAndQueue] error canceling jobs: %s", e)
+		} else if err != nil {
+			result := &api.CancellationResult{CancelledIds: cancelledIds}
+			return result, status.Errorf(codes.Unavailable, "[cancelJobsBySetAndQueue] error checking permissions: %s", err)
+		}
+		cancelledIds = append(cancelledIds, result.CancelledIds...)
 
-			if util.CloseToDeadline(ctx, time.Second*1) {
-				return &api.CancellationResult{CancelledIds: cancelledIds}, status.Errorf(
-					codes.DeadlineExceeded,
-					"not all jobs were cancelled: took too long")
-			}
+		// TODO I think the right way to do this is to include a timeout with the call to Redis
```

Comment:
I think we need to update to a newer version of go-redis with this supported, however that indeed would be better.